Genshi doesn't work on Python 3.11: code expected at least 18 arguments, got 16

Example of error when running tests:

```
======================================================================
FAIL: OldTextTemplate (genshi.template.text)
Doctest: genshi.template.text.OldTextTemplate
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vstinner/python/main/Lib/doctest.py", line 2216, in runTest
    raise self.failureException(self.format_failure(new.getvalue()))
AssertionError: Failed doctest test for genshi.template.text.OldTextTemplate
  File "/home/vstinner/dev/genshi/genshi/template/text.py", line 239, in OldTextTemplate

----------------------------------------------------------------------
File "/home/vstinner/dev/genshi/genshi/template/text.py", line 244, in genshi.template.text.OldTextTemplate
Failed example:
    tmpl = OldTextTemplate('''Dear $name,

    We have the following items for you:
    #for item in items
     * $item
    #end

    All the best,
    Foobar''')
Exception raised:
    Traceback (most recent call last):
      File "/home/vstinner/python/main/Lib/doctest.py", line 1348, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest genshi.template.text.OldTextTemplate[0]>", line 1, in <module>
        tmpl = OldTextTemplate('''Dear $name,
      File "/home/vstinner/dev/genshi/genshi/template/base.py", line 421, in __init__
        self._stream = self._parse(source, encoding)
      File "/home/vstinner/dev/genshi/genshi/template/text.py", line 293, in _parse
        for kind, data, pos in interpolate(text, self.filepath, lineno,
      File "/home/vstinner/dev/genshi/genshi/template/interpolation.py", line 77, in interpolate
        expr = Expression(chunk.strip(), pos[0], pos[1],
      File "/home/vstinner/dev/genshi/genshi/template/eval.py", line 71, in __init__
        self.code = _compile(node, self.source, mode=self.mode,
      File "/home/vstinner/dev/genshi/genshi/template/eval.py", line 453, in _compile
        return build_code_chunk(code, filename, name, lineno)
      File "/home/vstinner/dev/genshi/genshi/compat.py", line 110, in build_code_chunk
        return CodeType(*params)
    TypeError: code() argument 15 must be bytes, not tuple
----------------------------------------------------------------------
File "/home/vstinner/dev/genshi/genshi/template/text.py", line 253, in genshi.template.text.OldTextTemplate
Failed example:
    print(tmpl.generate(name='Joe', items=[1, 2, 3]).render(encoding=None))
Exception raised:
    Traceback (most recent call last):
      File "/home/vstinner/python/main/Lib/doctest.py", line 1348, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest genshi.template.text.OldTextTemplate[1]>", line 1, in <module>
        print(tmpl.generate(name='Joe', items=[1, 2, 3]).render(encoding=None))
    NameError: name 'tmpl' is not defined
```

--- a/genshi/compat.py
+++ b/genshi/compat.py
@@ -88,13 +88,33 @@ if IS_PYTHON2:
                         lineno, code.co_lnotab, (), ())
 else:
     def get_code_params(code):
-        params = [code.co_nlocals, code.co_kwonlyargcount, code.co_stacksize,
-                  code.co_flags, code.co_code, code.co_consts, code.co_names,
-                  code.co_varnames, code.co_filename, code.co_name,
-                  code.co_firstlineno, code.co_lnotab, (), ()]
+        params = []
         if hasattr(code, "co_posonlyargcount"):
-            # PEP 570 added "positional only arguments"
-            params.insert(1, code.co_posonlyargcount)
+            # PEP 570 -- Python Positional-Only Parameters
+            params.append(code.co_posonlyargcount)
+        params.extend([code.co_kwonlyargcount, code.co_nlocals,
+                       code.co_stacksize, code.co_flags, code.co_code,
+                       code.co_consts, code.co_names, code.co_varnames,
+                       code.co_filename, code.co_name])
+        if hasattr(code, "co_qualname"):
+            # https://bugs.python.org/issue13672
+            params.append(code.co_qualname)
+        params.append(code.co_firstlineno)
+        if hasattr(code, "co_linetable"):
+            # PEP 626 -- Precise line numbers for debugging and other tools.
+            params.append(code.co_linetable)
+        else:
+            params.append(code.co_lnotab)
+        if hasattr(code, "co_endlinetable"):
+            # PEP 657 -- Include Fine Grained Error Locations in Tracebacks
+            params.append(code.co_endlinetable)
+        if hasattr(code, "co_columntable"):
+            # PEP 657 -- Include Fine Grained Error Locations in Tracebacks
+            params.append(code.co_columntable)
+        if hasattr(code, "co_exceptiontable"):
+            # https://bugs.python.org/issue40222
+            params.append(code.co_exceptiontable)
+        params.extend([(), ()])
         return tuple(params)
 
 
@@ -106,15 +126,13 @@ else:
                 co_name=name,
                 co_firstlineno=lineno,
             )
-        params =  [0, code.co_nlocals, code.co_kwonlyargcount,
-                  code.co_stacksize, code.co_flags | 0x0040,
-                  code.co_code, code.co_consts, code.co_names,
-                  code.co_varnames, filename, name, lineno,
-                  code.co_lnotab, (), ()]
-        if hasattr(code, "co_posonlyargcount"):
-            # PEP 570 added "positional only arguments"
-            params.insert(2, code.co_posonlyargcount)
-        return CodeType(*params)
+        # XXX: This isn't equivalent to the above "replace" code (we overwrite
+        # co_argcount, co_flags, co_freevars, and co_cellvars). Should one of
+        # them be changed?
+        return CodeType(0, code.co_kwonlyargcount, code.co_nlocals,
+                        code.co_stacksize, code.co_flags | 0x0040, code.co_code,
+                        code.co_consts, code.co_names, code.co_varnames,
+                        filename, name, lineno, code.co_lnotab, (), ())
 
 
 # In Python 3.8, Str and Ellipsis was replaced by Constant

